REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [49]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [68]:
import csv

with open('data/businesses.csv', encoding='ISO-8859-1') as csvfile:
    reader = csv.reader(csvfile)
    number_records = len(list(reader))
    #print(number_records)
    print("Number of records:", number_records)

Number of records: 6316


#### Question 2b: How many unique business IDs are there?  

In [66]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
unique_ID = df['business_id'].nunique()
#print(unique_ID)
print('Number of Unique Business IDs:', unique_ID)




Number of Unique Business IDs: 6315


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [70]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
SF_businesses = df[df['city']=='San Francisco']
top_5_businesses = SF_businesses['name'].value_counts().head(5)
#print(top_5_businesses)
print('Top 5 Most Common Businesses:', top_5_businesses)

Top 5 Most Common Businesses: name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64


In [62]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
SF_businesses = df[df['city']=='San Francisco']
total_businesses = SF_businesses.shape[0]
#print(total_businesses)
print('Total number of businesses in SF:', total_businesses)

Total number of businesses in SF: 6315


## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [ ]:
#i am guessing they are stored as a string, but i will double check

In [71]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
col_name = 'postal_code'
data_type = df[col_name].dtype
#print(f"{data_type}")
print(f"postal code data type '{col_name}' is: {data_type}")

postal code data type 'postal_code' is: object


In [ ]:
#not sure what object means here. perhaps it's because there are NaN values?

#### Question 3b

What are the unique values of postal_code?

In [76]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
unique_postal_code = df['postal_code'].nunique()
#print(unique_postal_code)
print('Number of Unique Postal Codes:', unique_postal_code)

Number of Unique Postal Codes: 46


#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [73]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [77]:
df = df.dropna(subset=['postal_code'])
bus_valid = df[df['postal_code'].isin(validZip)]
number_bus = len(bus_valid)
#print(f"{number_bus}")
print(f"{number_bus} businesses in data frame 'bus_valid'.")

5999 businesses in data frame 'bus_valid'.


In [78]:
vb = bus[bus['postal_code'].isin(validZip)]


## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [ ]:
vb[pd.isnull(vb['longitude'])]
#i am a bit confused by this code here. is vb valid business? I don't understand why pd is nested like that
#is this code a way to condense and keep the code concise? 

In [3]:
import pandas as pd

In [5]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
df = df.dropna(subset=['postal_code'])
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

bus_valid = df[df['postal_code'].isin(validZip)]
missing_long = bus_valid['longitude'].isnull().sum()
print(missing_long)

2483


#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [30]:
import pandas as pd

In [31]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
df = df.dropna(subset=['postal_code'])
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [32]:
zip_df = pd.DataFrame({'postal_code': validZip})
null_lon_counts = []
not_null_lon_counts = []

In [33]:
for zipcode in valid_zip:
    zip_2 = df[df['postal_code'] == zipcode]
    null_lon_count = zip_2['longitude'].isnull().sum()
    not_null_lon_count = len(zip_2) - null_lon_count
    null_lon_counts.append(null_lon_count)
    not_null_lon_counts.append(not_null_lon_count)

zip_df['null_lon'] = null_lon_counts
zip_df['not_null_lon'] = not_null_lon_counts

print(zip_df)

   postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45


#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [37]:
df = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
df = df.dropna(subset=['postal_code'])
missing_lon = df[df['longitude'].isnull()]['postal_code'].value_counts(normalize=True)
print(missing_lon.head())

postal_code
94103    0.099921
94110    0.099524
94107    0.097938
94102    0.077716
94109    0.057891
Name: proportion, dtype: float64


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

SyntaxError: invalid syntax (448241739.py, line 1)

In [51]:
#alternativly, using the count function 
ins = pd.read_csv('data/inspections.csv')
business_counts = df['business_id'].value_counts()
unique_bus= df['business_id'].nunique()
if business_counts.max() == 1:
    print("one inspection")
else:
    print("multiple inspections")


one inspection


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [58]:
import pandas as pd
ins = pd.read_csv('data/inspections.csv')
type_counts = ins['type'].value_counts()
type_not_rare = type_counts[type_counts > 10].index
ins2 = ins[ins['type'].isin(type_not_rare)].copy()
print(ins2['type'].value_counts())

type
routine    15429
Name: count, dtype: int64


In [59]:
print(ins2['type'].value_counts())

type
routine    15429
Name: count, dtype: int64


#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [61]:
#ins = pd.read_csv('data/inspections.csv')
#ins['year'] = ins['date'].str[:4]
#print(ins.head())
#i am getting an error here because date is perhaps not in string form?
print(ins['date'].dtype)
#so the dates are actually in int64 and I need to convert to str?

int64


In [66]:
ins = pd.read_csv('data/inspections.csv')
ins['date'] = ins['date'].astype(str)
ins['year'] = ins['date'].str[:4]
print(ins.head())

   business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014


#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [68]:
import pandas as pd
ins = pd.read_csv('data/inspections.csv')
ins['date'] = ins['date'].astype(str)
ins['year'] = ins['date'].str[:4]
min_year = ins['year'].min()
max_year = ins['year'].max()
print(min_year, max_year)
#4 years total 

2013 2016


In [70]:
inspections_each_year = ins['year'].value_counts().sort_index()
print(inspections_each_year)
#here are the number of inspections each year from 2013 to 2016

year
2013      38
2014    5629
2015    3923
2016    5840
Name: count, dtype: int64


In [85]:
less_than_50_ins = inspections_each_year[inspections_each_year < 50].index
ins3 = ins[ins['year'].isin(less_than_50_ins)].copy()
print("ins3:", ins3.head())

ins3:       business_id  score      date     type  year
534           695     96  20131231  routine  2013
761          1043     98  20131226  routine  2013
804          1124     92  20131230  routine  2013
1034         1458     72  20131227  routine  2013
1680         2245     91  20131227  routine  2013


Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [86]:
ins_2016 = ins3[ins3['year'] == '2016']
print(ins_2016.head())
#no inspections in 2016???

Empty DataFrame
Columns: [business_id, score, date, type, year]
Index: []


In [87]:
print(ins3['year'].unique())
#does this mean that the only year with inspections is 2013?

['2013']


### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [89]:
business = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins_2016 = pd.read_csv('data/inspections.csv')
ins_2016['date'] = ins_2016['date'].astype(str)
#have to convert from int64 to str

In [93]:
ins_2016['year'] = ins_2016['date'].str[:4]
bus_ins_merge = pd.merge(business, ins_2016[ins_2016['year'] == '2016'],
                       on='business_id', how='left')

print(bus_ins_merge.head(10))

   business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   
5           45                 CHARLIE'S DELI CAFE   
6           45                 CHARLIE'S DELI CAFE   
7           48                          ART'S CAFE   
8           50                          SUSHI ZONE   
9           54                 RHODA GOLDMAN PLAZA   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4    

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [95]:
lowest_rated = bus_ins_merge.sort_values(by='score', ascending=True)
lowest_rated_20 = lowest_rated.head(20)
print(lowest_rated_20[['name', 'address', 'score']])
#here are the 20 lowest rated places

                             name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
7325                    POKI TIME     2101 LOMBARD ST    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
6938         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
3563    

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!